sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/joins.scala

```diff
@@ -152,3 +153,51 @@ object EliminateOuterJoin extends Rule[LogicalPlan] with PredicateHelper {
       if (j.joinType == newJoinType) f else Filter(condition, j.copy(joinType = newJoinType))
   }
 }
+
+/**
+ * Correctly handle PythonUDF which need access both side of join side by changing the new join
+ * type to Cross.
+ */
+object PullOutPythonUDFInJoinCondition extends Rule[LogicalPlan] with PredicateHelper {
+  def hasPythonUDF(expression: Expression): Boolean = {
+    expression.collectFirst { case udf: PythonUDF => udf }.isDefined
```

Comment:
I see no check that this requires attributes from both sides, shall we add it? I see that if this is not true the predicate should have been already pushed down, but an additional sanity check is worth IMHO

Reply:
This doesn't matter. We can't evaluate python udf in the join condition, and need to pull it out, that's all.

For python udf accessing attributes from only one side, these would be pushed down by other rules. If they don't (e.g. user disables filter pushdown rule), we need to pull them out here, too. Anyway it's orthogonal to this rule.

Reply:
thanks for the explanation @cloud-fan , makes sense, can we then update the comment of this class in order to reflect what it actually does?

Reply:
Thanks, done in 87f0f50.